File: ts/src/report.ts
/**
 * Validation issues and the report produced by a validation run.
 */

import { type Value, valueToString } from "./value.ts";

/** Severity of an {@link Issue}. */
export type Severity = "error" | "warning";

/** A single problem found while validating the environment. */
export interface Issue {
  readonly var: string;
  readonly severity: Severity;
  readonly message: string;
}

/** Construct an error-severity issue. */
export function errorIssue(variable: string, message: string): Issue {
  return { var: variable, severity: "error", message };
}

/** Construct a warning-severity issue. */
export function warningIssue(variable: string, message: string): Issue {
  return { var: variable, severity: "warning", message };
}

/** The JSON shape produced by {@link Report.toJSON}. Secrets are masked. */
export interface ReportJson {
  readonly ok: boolean;
  readonly errors: number;
  readonly warnings: number;
  readonly issues: ReadonlyArray<{
    readonly var: string;
    readonly severity: Severity;
    readonly message: string;
  }>;
  readonly resolved: Record<string, string>;
}

/** Mask used for any variable declared `secret = true`. */
const SECRET_MASK = "******";

/**
 * The outcome of validating a set of variables against a schema.
 *
 * `resolved` holds the successfully coerced, default-filled values keyed by
 * variable name — exactly the typed config you can hand to the rest of your
 * application. Secret values are never exposed in {@link Report.toText} or
 * {@link Report.toJSON}.
 */
export class Report {
  readonly issues: Issue[] = [];
  /** Successfully resolved values, keyed by variable name. Includes defaults. */
  readonly resolved = new Map<string, Value>();
  /** Names of variables whose schema marked them `secret = true`. */
  private readonly secrets = new Set<string>();

  /** Mark a variable as secret so its value is masked in all output. */
  markSecret(variable: string): void {
    this.secrets.add(variable);
  }

  /** Record an issue (error or warning). */
  add(issue: Issue): void {
    this.issues.push(issue);
  }

  /** Record a successfully resolved value. */
  resolve(variable: string, value: Value): void {
    this.resolved.set(variable, value);
  }

  /** True if any issue is an error. */
  hasErrors(): boolean {
    return this.issues.some((i) => i.severity === "error");
  }

  /** All error-severity issues, in insertion order. */
  errors(): Issue[] {
    return this.issues.filter((i) => i.severity === "error");
  }

  /** All warning-severity issues, in insertion order. */
  warnings(): Issue[] {
    return this.issues.filter((i) => i.severity === "warning");
  }

  /** Whether a variable was declared secret. */
  isSecret(variable: string): boolean {
    return this.secrets.has(variable);
  }

  private displayValue(variable: string, value: Value): string {
    return this.isSecret(variable) ? SECRET_MASK : valueToString(value);
  }

  /** Render a human-readable report. Secret values are masked. */
  toText(): string {
    let out = "";
    for (const issue of this.issues) {
      out += `${issue.severity}: ${issue.var}: ${issue.message}\n`;
    }
    const err = this.errors().length;
    const warn = this.warnings().length;
    if (err === 0 && warn === 0) {
      out += `ok: ${this.resolved.size} variable(s) validated\n`;
    } else {
      out += `${err} error(s), ${warn} warning(s)\n`;
    }
    return out;
  }

  /**
   * Render the report as a JSON-serialisable object. Secret values are masked.
   * `resolved` preserves the schema's variable order (insertion order).
   */
  toJSON(): ReportJson {
    const resolved: Record<string, string> = {};
    for (const [name, value] of this.resolved) {
      resolved[name] = this.displayValue(name, value);
    }
    return {
      ok: !this.hasErrors(),
      errors: this.errors().length,
      warnings: this.warnings().length,
      issues: this.issues.map((i) => ({
        var: i.var,
        severity: i.severity,
        message: i.message,
      })),
      resolved,
    };
  }
}

File: rust/src/report.rs
//! Validation issues and the report produced by a run.

use std::collections::BTreeMap;
use std::fmt;

use crate::value::Value;

/// Severity of an [`Issue`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Severity {
    Error,
    Warning,
}

impl fmt::Display for Severity {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Severity::Error => f.write_str("error"),
            Severity::Warning => f.write_str("warning"),
        }
    }
}

/// A single problem found while validating the environment.
#[derive(Debug, Clone, PartialEq)]
pub struct Issue {
    pub var: String,
    pub severity: Severity,
    pub message: String,
}

impl Issue {
    pub fn error(var: impl Into<String>, message: impl Into<String>) -> Self {
        Issue {
            var: var.into(),
            severity: Severity::Error,
            message: message.into(),
        }
    }

    pub fn warning(var: impl Into<String>, message: impl Into<String>) -> Self {
        Issue {
            var: var.into(),
            severity: Severity::Warning,
            message: message.into(),
        }
    }
}

/// The outcome of validating a set of variables against a schema.
#[derive(Debug, Default)]
pub struct Report {
    pub issues: Vec<Issue>,
    /// Successfully resolved values, keyed by variable name. Includes defaults.
    pub resolved: BTreeMap<String, Value>,
    /// Names of variables whose schema marked them `secret = true`.
    pub(crate) secrets: BTreeMap<String, ()>,
}

impl Report {
    pub fn has_errors(&self) -> bool {
        self.issues.iter().any(|i| i.severity == Severity::Error)
    }

    pub fn errors(&self) -> impl Iterator<Item = &Issue> {
        self.issues.iter().filter(|i| i.severity == Severity::Error)
    }

    pub fn warnings(&self) -> impl Iterator<Item = &Issue> {
        self.issues
            .iter()
            .filter(|i| i.severity == Severity::Warning)
    }

    pub(crate) fn is_secret(&self, var: &str) -> bool {
        self.secrets.contains_key(var)
    }

    fn display_value(&self, var: &str, value: &Value) -> String {
        if self.is_secret(var) {
            "******".to_string()
        } else {
            value.to_string()
        }
    }

    /// Render a human-readable report. Secret values are masked.
    pub fn to_text(&self) -> String {
        let mut out = String::new();
        for issue in &self.issues {
            out.push_str(&format!(
                "{}: {}: {}\n",
                issue.severity, issue.var, issue.message
            ));
        }
        let err = self.errors().count();
        let warn = self.warnings().count();
        if err == 0 && warn == 0 {
            out.push_str(&format!(
                "ok: {} variable(s) validated\n",
                self.resolved.len()
            ));
        } else {
            out.push_str(&format!("{err} error(s), {warn} warning(s)\n"));
        }
        out
    }

    /// Render the report as a JSON document. Secret values are masked.
    pub fn to_json(&self) -> serde_json::Value {
        let issues: Vec<_> = self
            .issues
            .iter()
            .map(|i| {
                serde_json::json!({
                    "var": i.var,
                    "severity": i.severity.to_string(),
                    "message": i.message,
                })
            })
            .collect();
        let resolved: serde_json::Map<String, serde_json::Value> = self
            .resolved
            .iter()
            .map(|(k, v)| {
                (
                    k.clone(),
                    serde_json::Value::String(self.display_value(k, v)),
                )
            })
            .collect();
        serde_json::json!({
            "ok": !self.has_errors(),
            "errors": self.errors().count(),
            "warnings": self.warnings().count(),
            "issues": issues,
            "resolved": resolved,
        })
    }
}
